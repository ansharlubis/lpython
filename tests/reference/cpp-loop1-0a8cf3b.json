{
    "basename": "cpp-loop1-0a8cf3b",
    "cmd": "lpython --no-color --show-cpp {infile}",
    "infile": "tests/loop1.py",
    "infile_hash": "c6799c859004650fdb3abca560c74e978e8f0d22d9f1e3466a074017",
    "outfile": null,
    "outfile_hash": null,
    "stdout": "cpp-loop1-0a8cf3b.stdout",
    "stdout_hash": "aac431590c9702b7f608115d729d71eba2a617a0c9230dc0c9d85d5b",
    "stderr": null,
    "stderr_hash": null,
    "returncode": 0
}