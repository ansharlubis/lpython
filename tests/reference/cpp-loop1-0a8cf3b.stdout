([1m[35mTranslationUnit[39m[0m 
   ([33mSymbolTable[39m 
      1 
      {
         main0: 
            ([1m[35mSubroutine[39m[0m 
               ([33mSymbolTable[39m 
                  5 
                  {
                     i: 
                        ([1m[35mVariable[39m[0m 
                           5 
                           i [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     j: 
                        ([1m[35mVariable[39m[0m 
                           5 
                           j [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 8 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               main0 [] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 5 [33mi[39m) 
                  ([1m[35mFunctionCall[39m[0m 1 [33mtest_factorial_1[39m () [
                     (([1m[35mIntegerConstant[39m[0m [36m4[39m 
                        ([1m[35mInteger[39m[0m 4 [])))] 
                     ([1m[35mInteger[39m[0m 4 []) () ()) ()) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 5 [33mi[39m) 
                  ([1m[35mFunctionCall[39m[0m 1 [33mtest_factorial_2[39m () [
                     (([1m[35mIntegerConstant[39m[0m [36m4[39m 
                        ([1m[35mInteger[39m[0m 4 [])))] 
                     ([1m[35mInteger[39m[0m 4 []) () ()) ()) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 5 [33mj[39m) 
                  ([1m[35mFunctionCall[39m[0m 1 [33mtest_factorial_3[39m () [
                     (([1m[35mIntegerConstant[39m[0m [36m5[39m 
                        ([1m[35mInteger[39m[0m 4 [])))] 
                     ([1m[35mInteger[39m[0m 8 []) () ()) ())] [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m () .false. .false.), 
         test_factorial_1: 
            ([1m[35mFunction[39m[0m 
               ([33mSymbolTable[39m 
                  2 
                  {
                     _lpython_return_variable: 
                        ([1m[35mVariable[39m[0m 
                           2 
                           _lpython_return_variable [1m[32m
                           ReturnVar[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     result: 
                        ([1m[35mVariable[39m[0m 
                           2 
                           result [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     x: 
                        ([1m[35mVariable[39m[0m 
                           2 
                           x [1m[32m
                           In[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               test_factorial_1 [
               ([1m[35mVar[39m[0m 2 [33mx[39m)] [
               ([1m[35mIf[39m[0m 
                  ([1m[35mIntegerCompare[39m[0m 
                     ([1m[35mVar[39m[0m 2 [33mx[39m) [1m[32m
                     Lt[39m[0m ([1m[35mIntegerConstant[39m[0m [36m0[39m 
                     ([1m[35mInteger[39m[0m 4 [])) 
                     ([1m[35mLogical[39m[0m 4 []) ()) [
                  ([1m[35m=[39m[0m 
                     ([1m[35mVar[39m[0m 2 [33m_lpython_return_variable[39m) ([1m[35mIntegerConstant[39m[0m [36m0[39m 
                     ([1m[35mInteger[39m[0m 4 [])) ()) 
                  ([1m[35mReturn[39m[0m)] []) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 2 [33mresult[39m) ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                  ([1m[35mInteger[39m[0m 4 [])) ()) 
               ([1m[35mWhileLoop[39m[0m 
                  ([1m[35mIntegerCompare[39m[0m 
                     ([1m[35mVar[39m[0m 2 [33mx[39m) [1m[32m
                     Gt[39m[0m ([1m[35mIntegerConstant[39m[0m [36m0[39m 
                     ([1m[35mInteger[39m[0m 4 [])) 
                     ([1m[35mLogical[39m[0m 4 []) ()) [
                  ([1m[35m=[39m[0m 
                     ([1m[35mVar[39m[0m 2 [33mresult[39m) 
                     ([1m[35mIntegerBinOp[39m[0m 
                        ([1m[35mVar[39m[0m 2 [33mresult[39m) [1m[32m
                        Mul[39m[0m 
                        ([1m[35mVar[39m[0m 2 [33mx[39m) 
                        ([1m[35mInteger[39m[0m 4 []) ()) ()) 
                  ([1m[35m=[39m[0m 
                     ([1m[35mVar[39m[0m 2 [33mx[39m) 
                     ([1m[35mIntegerBinOp[39m[0m 
                        ([1m[35mVar[39m[0m 2 [33mx[39m) [1m[32m
                        Sub[39m[0m ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                        ([1m[35mInteger[39m[0m 4 [])) 
                        ([1m[35mInteger[39m[0m 4 []) ()) ())]) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 2 [33m_lpython_return_variable[39m) 
                  ([1m[35mVar[39m[0m 2 [33mresult[39m) ()) 
               ([1m[35mReturn[39m[0m)] 
               ([1m[35mVar[39m[0m 2 [33m_lpython_return_variable[39m) [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m ()), 
         test_factorial_2: 
            ([1m[35mFunction[39m[0m 
               ([33mSymbolTable[39m 
                  3 
                  {
                     _lpython_return_variable: 
                        ([1m[35mVariable[39m[0m 
                           3 
                           _lpython_return_variable [1m[32m
                           ReturnVar[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     i: 
                        ([1m[35mVariable[39m[0m 
                           3 
                           i [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     result: 
                        ([1m[35mVariable[39m[0m 
                           3 
                           result [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     x: 
                        ([1m[35mVariable[39m[0m 
                           3 
                           x [1m[32m
                           In[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               test_factorial_2 [
               ([1m[35mVar[39m[0m 3 [33mx[39m)] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 3 [33mresult[39m) ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                  ([1m[35mInteger[39m[0m 4 [])) ()) 
               ([1m[35mDoLoop[39m[0m 
                  (
                     ([1m[35mVar[39m[0m 3 [33mi[39m) ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                     ([1m[35mInteger[39m[0m 4 [])) 
                     ([1m[35mIntegerBinOp[39m[0m 
                        ([1m[35mIntegerBinOp[39m[0m 
                           ([1m[35mVar[39m[0m 3 [33mx[39m) [1m[32m
                           Add[39m[0m ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                           ([1m[35mInteger[39m[0m 4 [])) 
                           ([1m[35mInteger[39m[0m 4 []) ()) [1m[32m
                        Sub[39m[0m ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                        ([1m[35mInteger[39m[0m 4 [])) 
                        ([1m[35mInteger[39m[0m 4 []) ()) ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                     ([1m[35mInteger[39m[0m 4 []))) [
                  ([1m[35m=[39m[0m 
                     ([1m[35mVar[39m[0m 3 [33mresult[39m) 
                     ([1m[35mIntegerBinOp[39m[0m 
                        ([1m[35mVar[39m[0m 3 [33mresult[39m) [1m[32m
                        Mul[39m[0m 
                        ([1m[35mVar[39m[0m 3 [33mi[39m) 
                        ([1m[35mInteger[39m[0m 4 []) ()) ())]) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 3 [33m_lpython_return_variable[39m) 
                  ([1m[35mVar[39m[0m 3 [33mresult[39m) ()) 
               ([1m[35mReturn[39m[0m)] 
               ([1m[35mVar[39m[0m 3 [33m_lpython_return_variable[39m) [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m ()), 
         test_factorial_3: 
            ([1m[35mFunction[39m[0m 
               ([33mSymbolTable[39m 
                  4 
                  {
                     _lpython_return_variable: 
                        ([1m[35mVariable[39m[0m 
                           4 
                           _lpython_return_variable [1m[32m
                           ReturnVar[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 8 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     result: 
                        ([1m[35mVariable[39m[0m 
                           4 
                           result [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 8 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     x: 
                        ([1m[35mVariable[39m[0m 
                           4 
                           x [1m[32m
                           In[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               test_factorial_3 [
               ([1m[35mVar[39m[0m 4 [33mx[39m)] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 4 [33mresult[39m) 
                  ([1m[35mCast[39m[0m ([1m[35mIntegerConstant[39m[0m [36m0[39m 
                     ([1m[35mInteger[39m[0m 4 [])) [1m[32m
                     IntegerToInteger[39m[0m 
                     ([1m[35mInteger[39m[0m 8 []) ()) ()) 
               ([1m[35mIf[39m[0m 
                  ([1m[35mIntegerCompare[39m[0m 
                     ([1m[35mVar[39m[0m 4 [33mx[39m) [1m[32m
                     Lt[39m[0m ([1m[35mIntegerConstant[39m[0m [36m0[39m 
                     ([1m[35mInteger[39m[0m 4 [])) 
                     ([1m[35mLogical[39m[0m 4 []) ()) [
                  ([1m[35m=[39m[0m 
                     ([1m[35mVar[39m[0m 4 [33m_lpython_return_variable[39m) 
                     ([1m[35mVar[39m[0m 4 [33mresult[39m) ()) 
                  ([1m[35mReturn[39m[0m)] []) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 4 [33mresult[39m) 
                  ([1m[35mCast[39m[0m ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                     ([1m[35mInteger[39m[0m 4 [])) [1m[32m
                     IntegerToInteger[39m[0m 
                     ([1m[35mInteger[39m[0m 8 []) ()) ()) 
               ([1m[35mWhileLoop[39m[0m 
                  ([1m[35mIntegerCompare[39m[0m 
                     ([1m[35mVar[39m[0m 4 [33mx[39m) [1m[32m
                     Gt[39m[0m ([1m[35mIntegerConstant[39m[0m [36m0[39m 
                     ([1m[35mInteger[39m[0m 4 [])) 
                     ([1m[35mLogical[39m[0m 4 []) ()) [
                  ([1m[35m=[39m[0m 
                     ([1m[35mVar[39m[0m 4 [33mresult[39m) 
                     ([1m[35mIntegerBinOp[39m[0m 
                        ([1m[35mVar[39m[0m 4 [33mresult[39m) [1m[32m
                        Mul[39m[0m 
                        ([1m[35mCast[39m[0m 
                           ([1m[35mVar[39m[0m 4 [33mx[39m) [1m[32m
                           IntegerToInteger[39m[0m 
                           ([1m[35mInteger[39m[0m 8 []) ()) 
                        ([1m[35mInteger[39m[0m 8 []) ()) ()) 
                  ([1m[35m=[39m[0m 
                     ([1m[35mVar[39m[0m 4 [33mx[39m) 
                     ([1m[35mIntegerBinOp[39m[0m 
                        ([1m[35mVar[39m[0m 4 [33mx[39m) [1m[32m
                        Sub[39m[0m ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                        ([1m[35mInteger[39m[0m 4 [])) 
                        ([1m[35mInteger[39m[0m 4 []) ()) ())]) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 4 [33m_lpython_return_variable[39m) 
                  ([1m[35mVar[39m[0m 4 [33mresult[39m) ()) 
               ([1m[35mReturn[39m[0m)] 
               ([1m[35mVar[39m[0m 4 [33m_lpython_return_variable[39m) [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m ())
      }) 
   [
   ([1m[35mSubroutineCall[39m[0m 1 [33mmain0[39m () [] ())])
#include <iostream>
#include <string>
#include <vector>
#include <cassert>
#include <cmath>
#include <complex>
#include <Kokkos_Core.hpp>
#include <lfortran_intrinsics.h>

template <typename T>
Kokkos::View<T*> from_std_vector(const std::vector<T> &v)
{
    Kokkos::View<T*> r("r", v.size());
    for (size_t i=0; i < v.size(); i++) {
        r(i) = v[i];
    }
    return r;
}

// Forward declarations
void _lpython_main_program();
void main0();
int32_t test_factorial_1(int x);
int32_t test_factorial_2(int x);
int64_t test_factorial_3(int x);
namespace {
}

// Implementations
void _lpython_main_program()
{
    main0();
}

void main0()
{
    int i;
    long long j;
    i = test_factorial_1(4);
    i = test_factorial_2(4);
    j = test_factorial_3(5);
}

int32_t test_factorial_1(int x)
{
    int _lpython_return_variable;
    int result;
    if (x < 0) {
        _lpython_return_variable = 0;
        return _lpython_return_variable;
    }
    result = 1;
    while (x > 0) {
        result = result*x;
        x = x - 1;
    }
    _lpython_return_variable = result;
    return _lpython_return_variable;
}

int32_t test_factorial_2(int x)
{
    int _lpython_return_variable;
    int i;
    int result;
    result = 1;
    for (i=1; i<=x + 1 - 1; i++) {
        result = result*i;
    }
    _lpython_return_variable = result;
    return _lpython_return_variable;
}

int64_t test_factorial_3(int x)
{
    long long _lpython_return_variable;
    long long result;
    result = 0;
    if (x < 0) {
        _lpython_return_variable = result;
        return _lpython_return_variable;
    }
    result = 1;
    while (x > 0) {
        result = result*x;
        x = x - 1;
    }
    _lpython_return_variable = result;
    return _lpython_return_variable;
}

namespace {

void main2() {
    _lpython_main_program();
}

}
int main(int argc, char* argv[])
{
    Kokkos::initialize(argc, argv);
    main2();
    Kokkos::finalize();
    return 0;
}
